src/rtc.c

```diff
@@ -0,0 +1,518 @@
+ *
+ * @param[in]   block
+ *              Which RTC block to write to (0-2)
+ *
+ * @param[out]  data
+ *              RTC block data to write
+ *
+ * @return the busy status of the real-time clock
+ */
+uint8_t joybus_rtc_write( uint8_t block, const uint64_t * data )
+{
+    assert(block <= 2);
+
+    static uint64_t input[8] =
+    {
+        0x000000000A010800,
+        0xffffffffffffffff,
+        0xfffe000000000000,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    input[0] = (input[0] & 0xFFFFFFFFFFFFFF00) | block;
+    input[1] = *data;
+
+    controller_exec_pif( input, output );
+
+    return output[2] >> 56;
+}
+
+/**
+ * @brief Read the status of the real-time clock.
+ *
+ * Some flash carts require the RTC to be explicitly enabled.
+ * Some emulators and flash carts do not support RTC at all.
+ *
+ * @return whether the Joybus RTC is present on the cartridge.
+ */
+bool joybus_rtc_is_present( void )
+{
+    return (joybus_rtc_status() >> 8) == JOYBUS_RTC_IDENTIFIER;
+}
+
+/**
+ * @brief Read the status of the real-time clock.
+ *
+ * The RTC will be busy when the control data is in
+ * #JOYBUS_RTC_CONTROL_MODE_SET and is ready to be written to.
+ *
+ * @return whether the Joybus RTC status is busy
+ */
+bool joybus_rtc_is_busy( void )
+{
+    return (joybus_rtc_status() & 0xFF) == JOYBUS_RTC_STATUS_BUSY;
+}
+
+/**
+ * @brief Read the control block from the real-time clock.
+ *
+ * The control data should generally be one of the pre-defined modes:
+ * #JOYBUS_RTC_CONTROL_MODE_SET when writing to the RTC.
+ * #JOYBUS_RTC_CONTROL_MODE_RUN when not writing to the RTC.
+ *
+ * The calibration data is an opaque value that should be propagated when
+ * calling #joybus_rtc_write_control. Emulators and flash carts may return
+ * zeroes for this data, which is also fine. Based on reverse-engineering
+ * the real hardware, this data is simply described as "magical", but
+ * should be preserved in the interest of high-accuracy and compatibility.
+ *
+ * Generally, you should not need to call this function directly. Prefer
+ * calling the high-level #joybus_rtc_run and #joybus_rtc_set functions which
+ * handle the delays and calibration propagation. Call #joybus_rtc_status or
+ * #joybus_rtc_is_busy to determine which control mode the RTC is in.
+ *
+ * @param[out]  control
+ *              Destination pointer for the RTC control data.
+ *
+ * @param[out]  calibration
+ *              Destination pointer for the RTC calibration data.
+ */
+void joybus_rtc_read_control( uint16_t * control, uint32_t * calibration )
+{
+    uint64_t data;
+    joybus_rtc_read( 0, &data );
+
+    if( control != NULL ) *control = data >> 48;
+    if( calibration != NULL ) *calibration = data;
+}
+
+/**
+ * @brief Read the current date/time from the real-time clock.
+ *
+ * Your code should call this once per frame to update the rtc_time_t
+ * data structure.
```

Comment:
Once per frame seems overkill to me... what's the timing of this function?

Anyway I think this should be called once at boot and then an internal time should be calculated using timer_ticks(). This is why I suggest hooking this up to newlib, so that we can provide the correct abstraction to calculate real time much faster than polling joybus

Reply:
That recommendation is based off of what Animal Forest does. 🤷🏻

Reply:
Still, I think it's not a good suggestion at least for now. 

First, we don't expose an API to be able to fetch RTC in background. I don't know timing for `joybus_rtc_get`, but you need to wait for a full joybus handshake so it can't be very fast. I'm sure Animal Forest was using background fetching (issue the command, then wait for interrupt and process it). Once this library will be threading compatibly, maybe it'd make more sense.

Second, LibUltra doesn't have a 64-bit timer. We do expose a 64-bit timer via `timer_ticks()`. So with LibUltra, you have no way to keep up with the current time, because the 32-bit COP0 counter overflows too often. What I would expect is that our `gettimeofday` implementation would transparently read RTC once (at boot or first call), and then store the current time as a 64-bit offset, so that later it is sufficient to add that offset to `timer_ticks()` to obtain the correct time, and then use newlib functions to convert it into calendar dates when required.